File: cpp/D5.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <regex>
#include <unordered_set>
#include "time.h"
#include "text.h"

using namespace std;

vector<string> read_input() {
    ifstream input("inputs\\D5.input");
    return split_lines(&input);
}

int binary(string s) {
    string bin;
    for (const char c : s) {
        if (c == 'F' || c == 'L') bin += '0';
        if (c == 'B' || c == 'R') bin += '1';
    }
    return stoi(bin, nullptr, 2);
}

int seat_id(string seat) { 
    int row = binary(seat.substr(0,7));
    int col = binary(seat.substr(7));
    return row * 8 + col; 
}

int P1(vector<string> seats) {
    int max = 0;
    for (string seat : seats) {
        int id = seat_id(seat);
        if (id > max) max = id;
    }
    return max;
}

int P2(vector<string> seats) {
    unordered_set<int> ids;
    int min = INT_MAX;
    int max = INT_MIN;
    for (string seat : seats) {
        const int id = seat_id(seat);
        if (id < min) min = id;
        if (id > max) max = id;
        ids.insert(id);
    }

    for (int i = min; i <= max; i++) {
        if (ids.find(i) == ids.end())
            return i;
    }

    return -1;
}

int main() {
    run_with_time<int>([](void) { 
        return P1(read_input()); 
    });

    run_with_time<int>([](void) { 
        return P2(read_input()); 
    });
}

File: py/D5.py

from os import replace


def parseInput():
    file = open("../inputs/D5.input", "r")
    seats = []
    contents = file.read()
    for line in contents.splitlines():
        seats.append(line)
    return seats


def binary(s: str):
    bin = (s.replace("F", "0")
           .replace("B", "1")
           .replace("L", "0")
           .replace("R", "1"))
    return int(bin, 2)

def seatId(seat: str):
    row = binary(seat[0:7])
    col = binary(seat[7:])
    return row * 8 + col

def p1():
    return max(map(seatId, parseInput()))

def p2():
    seatIds = sorted(map(seatId, parseInput()))
    for i in range(len(seatIds) - 1):
        if seatIds[i] + 1 != seatIds[i + 1]:
            return seatIds[i] + 1


print(p1())
print(p2())
